### PR TITLE
Virtual hmd

### PR DESCRIPTION
**Pull request by HipsterSloth**
As a follow up to VirtualHMDs we now also support VirtualControllers: Tracking a single colored ping-pong ball and pretending it's a tracked controller. As with the virtual HMD there is no orientation and no buttons. The psnavi can also be parented to a virtual controller allowing for a tracked PSNavi (minus orientation). To add one or more virtual controllers do the following:

1. Edit `%appdata%\PSMoveService\ControllerManagerConfig.json`
1. Set "virtual_controller_count" to some value > 0
1. Go to the Controller Settings panel of PSMoveConfigTool
1. Launch PSMoveService once and then shut it down to create controller config files
1. Edit `%appdata%\PSMoveService\VirtualController_N.json`, where N is the virtual controller number
1. Set the "bulb_radius" to the appropriate size (2cm for a regulation ping-pong ball)
1. Re-launch PSMoveService and then PSMoveConfigTool
1. Set the Tracking Color on each virtual controller
1. Configure the color filtering on each virtual controller
1. Test the virtual controller tracking

### CODE REVIEW

**Comment by HipsterSloth** [comment]
@zelmon64 When you get a chance would you mind giving this a quick test? I made some pre-built release binaries so you don't have to sync and build.

[PSMoveService_0.9_alpha8.5.0.zip](https://github.com/cboulay/PSMoveService/files/1063139/PSMoveService_0.9_alpha8.5.0.zip)

**Comment by zelmon64** [comment]
@HipsterSloth this is great work!

So far using one virtual controller seems fine but two crashes with the log below. This happens both with and without a virtual HMD. I'll try to find the cause.

When you add the virtual controllers to PSMoveSteamVRBridge would it be possible to have them register as Vive tracker-like objects?

[2017-06-09 10:29:03.398]: main - Starting PSMoveService v0.9-alpha 8.5.0
[2017-06-09 10:29:03.406]: USBAsyncRequestManager::startup - Requested LibUSBApi
[2017-06-09 10:29:03.408]: USBAsyncRequestManager::startup - Creating LibUSBApi
[2017-06-09 10:29:03.411]: USBAsyncRequestManager::startup - Initialized USB API
[2017-06-09 10:29:03.415]: DeviceManager::startup - Platform Hotplug API is ENABLED
[2017-06-09 10:29:03.422]: bluetooth_get_host_address - Found a bluetooth radio
[2017-06-09 10:29:03.424]: bluetooth_get_host_address - Retrieved radio info
[2017-06-09 10:29:03.570]: VirtualController::open - Opening VirtualController(VirtualController_0).
[2017-06-09 10:29:03.575]: DeviceTypeManager::update_connected_devices - Device device_id 0 (UNKNOWN) opened

File: src/psmoveservice/Device/Enumerator/VirtualControllerEnumerator.cpp [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+int VirtualControllerEnumerator::virtual_controller_count= 0;
+
+// -- VirtualControllerDeviceEnumerator -----
+VirtualControllerEnumerator::VirtualControllerEnumerator()
+    : DeviceEnumerator(CommonDeviceState::VirtualController)
+{
+	m_deviceType= CommonDeviceState::VirtualController;
+    m_device_index= 0;
+
+    m_current_device_identifier= "VirtualController_0";
```
Comment (zelmon64) [inline src/psmoveservice/Device/Enumerator/VirtualControllerEnumerator.cpp]: It looks like all virtual controllers are being identified as "VirtualController_0" which might be the problem when trying to use more than one.

Comment (HipsterSloth) [inline src/psmoveservice/Device/Enumerator/VirtualControllerEnumerator.cpp]: Looks like I had an incorrect format specifier in VirtualControllerEnumerator::next()

https://github.com/cboulay/PSMoveService/pull/430/files/d600c7e10954cf81914de4df63e1b8afd8f0cabc#diff-a997644b7b062f7f3b123ada114ce214R50

**Comment by HipsterSloth** [comment]
> When you add the virtual controllers to PSMoveSteamVRBridge would it be possible to have them register as Vive tracker-like objects?

Yeah that's the plan, I was thinking of having them show up as a special kind of controller (separate from psmove controllers). Probably just use the standard default vive controller model.

**Comment by zelmon64** [comment]
@HipsterSloth that'd be great. I thought you probably were planning to but wanted to make sure.

The fix works fine (I managed to get my build environment running again after what felt like a few dozen reinstalls) but I noticed another little error. DeviceTypeManager calls the Virtual Controller "UNKNOWN".

```
[2017-06-10 09:38:18.31]: VirtualController::open - Opening VirtualController(VirtualController_0).
[2017-06-10 09:38:18.36]: DeviceTypeManager::update_connected_devices - Device device_id 0 (UNKNOWN) opened
[2017-06-10 09:38:18.41]: VirtualController::open - Opening VirtualController(VirtualController_1).
[2017-06-10 09:38:18.47]: DeviceTypeManager::update_connected_devices - Device device_id 1 (UNKNOWN) opened
```

**Pull request merged**